runtime/compiler/z/codegen/S390PrivateLinkage.cpp

```diff
@@ -1871,41 +1871,32 @@ J9::Z::PrivateLinkage::buildVirtualDispatch(TR::Node * callNode, TR::RegisterDep
          }
       else
          {
-         cursor =
-            generateRXInstruction(cg(), TR::InstOpCode::getExtendedLoadOpCode(), callNode, RegRA,
-                                   generateS390MemoryReference(classReg, offset, cg()));
-
+          // If unresolved/AOT, this instruction will be patched by _virtualUnresolvedHelper() with the correct offset
+          cursor = generateRXInstruction(cg(), TR::InstOpCode::getExtendedLoadOpCode(), callNode, RegRA,
+                                         generateS390MemoryReference(classReg, offset, cg()));
          if (unresolvedSnippet)
             {
             ((TR::S390VirtualUnresolvedSnippet *)unresolvedSnippet)->setPatchVftInstruction(cursor);
             }
 
-         /* A load immediate (LHI/LGHI) instruction MUST be generated here for all virtual calls because
-          *
-          * 1. picbuilder looks for this load immediate (and an LG before it) and patches in the correct offset at
-          * runtime for unresolved virtual calls
-          *
-          * 2. for all virtual calls, resolved or not, this load immediate is used by the VM to find where
-          * the method is in the VFT table when doing J2I transitions. We need this negative offset in a register
-          * because S390 load address instruction used to find VFT table entires can't handle negative offsets.
-          * (VFT table precedes the J9Class in memory; as a result of this, all VFT entries are some negative offsets
-          * away from the J9Class)
+         /* A load immediate (LHI/LGFI) instruction MUST be generated here for all virtual calls because
+          *    1. PicBuilder patches this instruction to the correct offset for unresolved/AOT cases
+          *    2. The VM expects R0 to have the VFT offset for J2I transitions
+          * For unresolved/AOT targets we must assume the worse case (offset can't fit in 16bits)
           *
+          * We need this negative offset in a register because the "LA" instruction used by the VM to find
+          * VFT table entries can't handle negative offsets
           *
-          * In fact we should not be using APIs (such as genLoadLongConstant) at all and be explicitly generating
-          * LHI/LGHI instruction. Similar occurrence below.
+          * A VFT offset of 0 means unresolved/AOT, otherwise the offset is always negative.
           */
-         if (offset >= (-32768))
+         if (offset >= (-32768) && offset < 0) // Offset fits in 16bits
```

Comment:
Can we use `MIN_IMMEDIATE_VAL` here instead of the hardcoded constant?

Reply:
Done